x-pack/elastic-agent/pkg/composable/providers/kubernetes/config.go

```diff
@@ -9,24 +9,72 @@ package kubernetes
 
 import (
 	"time"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
 )
 
 // Config for kubernetes provider
 type Config struct {
+	Scope     string    `config:"scope"`
+	Resources Resources `config:"resources"`
+
+	// expand config settings at root level of config too
 	KubeConfig     string        `config:"kube_config"`
+	Namespace      string        `config:"namespace"`
 	SyncPeriod     time.Duration `config:"sync_period"`
 	CleanupTimeout time.Duration `config:"cleanup_timeout" validate:"positive"`
 
-	// Needed when resource is a pod
+	// Needed when resource is a Pod or Node
 	Node string `config:"node"`
+}
 
-	// Scope of the provider (cluster or node)
-	Scope string `config:"scope"`
+// Resources config section for resources' config blocks
+type Resources struct {
+	Pod     *ResourceConfig `config:"pod"`
+	Node    *ResourceConfig `config:"node"`
+	Service *ResourceConfig `config:"service"`
+}
+
+// ResourceConfig for kubernetes resource
+type ResourceConfig struct {
+	KubeConfig     string        `config:"kube_config"`
+	Namespace      string        `config:"namespace"`
+	SyncPeriod     time.Duration `config:"sync_period"`
+	CleanupTimeout time.Duration `config:"cleanup_timeout" validate:"positive"`
+
+	// Needed when resource is a Pod or Node
+	Node string `config:"node"`
 }
 
 // InitDefaults initializes the default values for the config.
 func (c *Config) InitDefaults() {
-	c.SyncPeriod = 10 * time.Minute
 	c.CleanupTimeout = 60 * time.Second
+	c.SyncPeriod = 10 * time.Minute
 	c.Scope = "node"
 }
+
+// Validate ensures correctness of config
+func (c *Config) Validate() error {
+	// Check if resource is service. If yes then default the scope to "cluster".
+	if c.Resources.Service != nil {
+		if c.Scope == "node" {
+			logp.L().Warnf("can not set scope to `node` when using resource `Service`. resetting scope to `cluster`")
+		}
+		c.Scope = "cluster"
+	}
+	baseCfg := &ResourceConfig{
+		CleanupTimeout: c.CleanupTimeout,
+		SyncPeriod:     c.CleanupTimeout,
+		KubeConfig:     c.KubeConfig,
+		Namespace:      c.Namespace,
+		Node:           c.Node,
+	}
+	if c.Resources.Pod == nil {
+		c.Resources.Pod = baseCfg
```

Comment:
what if the user only overrides `resources.pod.namespace`? Does that mean that the rest of settings will be empty?

Reply:
See https://github.com/elastic/beats/pull/26801#discussion_r673780920